reconcile/gitlab_owners.py

```diff
@@ -0,0 +1,270 @@
+import base64
+import json
+import logging
+import pathlib
+
+from dateutil import parser as dateparser
+
+from ruamel import yaml
+
+from reconcile import queries
+from utils.gitlab_api import GitLabApi
+from utils.gitlab_api import MRState
+
+
+QONTRACT_INTEGRATION = 'gitlab-owners'
+
+APPROVAL_LABEL = 'approved'
+
+_LOG = logging.getLogger(__name__)
+
+
+class ProjectOwners:
+    """
+    Abstracts the owners of a project with per-path granularity.
+    """
+
+    def __init__(self, gitlab_client):
+        self._gitlab = gitlab_client
+        self._owners_map = self._get_owners_map()
+
+    def get_all_owners(self, path):
+        """
+        Gets all the owners of a given path, no matter in which
+        level of the filesystem tree the owner was specified.
+        """
+        path_owners = set()
+        for owned_path, owners in self._owners_map.items():
+            if path.startswith(owned_path):
+                path_owners.update(owners)
+        if not path_owners:
+            raise KeyError(f'No owners for path {path!r}')
+        # Returns a sorted list of unique owners
+        return sorted(path_owners)
+
+    def get_closest_owners(self, path):
+        """
+        Gets all closest owners of a given path, no matter in which
+        level of the filesystem tree the owner was specified.
+        """
+        candidates = []
+
+        for owned_path in self._owners_map:
+            if path.startswith(owned_path):
+                candidates.append(owned_path)
+
+        if not candidates:
+            raise KeyError(f'No owners for path {path!r}')
+
+        # The longest owned_path is the chosen
+        elected = max(candidates, key=lambda x: len(x))
+        # Returns a sorted list of unique owners
+        return sorted(set(self._owners_map[elected]))
+
+    def _get_owners_map(self):
+        """
+        Maps all the OWNERS files content to their respective
+        owned directory.
+        """
+        owners_map = dict()
+        repo_tree = self._gitlab.get_repository_tree(ref='master')
+        for item in repo_tree:
+            if item['name'] != 'OWNERS':
```

Comment:
what happens in the case of a change to the OWNERS file?

Reply:
is this even the right place to ask? :)

Reply:
in each integration loop we get the current owners from master...